sql-plugin/src/main/scala/com/nvidia/spark/rapids/AbstractGpuJoinIterator.scala

```diff
@@ -0,0 +1,311 @@
+      }
+
+      if (ret.isDefined) {
+        // We are about to return something. We got everything we need from it so now let it spill
+        // if there is more to be gathered later on.
+        gathererStore.foreach(_.allowSpilling())
+      }
+      ret
+    }
+  }
+}
+
+/**
+ * Base class for join iterators that split and spill batches to avoid GPU OOM errors.
+ * @param gatherNvtxName name to use for the NVTX range when producing the join gather maps
+ * @param stream iterator to produce the batches for the streaming side input of the join
+ * @param streamAttributes attributes corresponding to the streaming side input
+ * @param builtBatch batch for the built side input of the join
+ * @param targetSize configured target batch size in bytes
+ * @param spillCallback callback to use when spilling
+ * @param joinTime metric to record GPU time spent in join
+ * @param streamTime metric to record time spent producing streaming side batches
+ * @param totalTime metric to record total time in the iterator
+ */
+abstract class SplittableJoinIterator(
+    gatherNvtxName: String,
+    stream: Iterator[LazySpillableColumnarBatch],
+    streamAttributes: Seq[Attribute],
+    builtBatch: LazySpillableColumnarBatch,
+    targetSize: Long,
+    spillCallback: SpillCallback,
+    joinTime: GpuMetric,
+    streamTime: GpuMetric,
+    totalTime: GpuMetric)
+    extends AbstractGpuJoinIterator(
+      gatherNvtxName,
+      targetSize,
+      joinTime = joinTime,
+      totalTime = totalTime) with Logging {
+  // For some join types even if there is no stream data we might output something
+  private var isInitialJoin = true
+  // If the join explodes this holds batches from the stream side split into smaller pieces.
+  private val pendingSplits = mutable.Queue[SpillableColumnarBatch]()
+
+  protected def computeNumJoinRows(cb: ColumnarBatch): Long
+
+  /**
+   * Create a join gatherer.
+   * @param cb next column batch from the streaming side of the join
+   * @param numJoinRows if present, the number of join output rows computed for this batch
+   * @return some gatherer to use next or None if there is no next gatherer or the loop should try
+   *         to build the gatherer again (e.g.: to skip a degenerate join result batch)
+   */
+  protected def createGatherer(cb: ColumnarBatch, numJoinRows: Option[Long]): Option[JoinGatherer]
+
+  override def hasNextStreamBatch: Boolean = {
+    isInitialJoin || pendingSplits.nonEmpty || stream.hasNext
+  }
+
+  override def setupNextGatherer(startNanoTime: Long): Option[JoinGatherer] = {
+    val wasInitialJoin = isInitialJoin
+    isInitialJoin = false
+    if (pendingSplits.nonEmpty || stream.hasNext) {
+      val cb = if (pendingSplits.nonEmpty) {
+        withResource(pendingSplits.dequeue()) {
+          _.getColumnarBatch()
+        }
+      } else {
+        val batch = withResource(stream.next()) { lazyBatch =>
+          // TODO: Worth having a releaseBatch method for LazySpillableBatch?
```

Comment:
Are you planning on doing this?

Reply:
I left this here to see if you thought that would be a good idea or not.  Seems a bit wasteful to walk through a column batch to increment refcounts only to turn around and decrement them all immediately again.  But it's probably not a significant performance problem either.  If you want me to add this here or part of a followup, happy to do so.

Reply:
I thought it might have been. For me it is all about cleanliness of the code, because like you said the performance difference is minimal. I think a `release` would make it simpler to follow what is happening.

Reply:
I added a `batchRelease` method to lazy spillable.  I made it act semantically like a "get batch and close without closing the resulting batch".  That means for spill-only it does _not_ end up closing the underlying resource, since that seems to be the whole point of that class.  Let me know if this makes sense, or if the semantics are more confusing than the incref approach that was there before.